python/ray/serve/_private/publish_provider.py

```diff
@@ -0,0 +1,73 @@
+from abc import ABC, abstractmethod
+from subprocess import CalledProcessError, check_output
+from tempfile import NamedTemporaryFile
+from typing import Dict, Optional
+
+import click
+import yaml
+
+from ray._private.utils import import_attr
+
+# Method that dynamically imported modules must implement to return a `PublishProvider`.
+PUBLISH_PROVIDER_FACTORY_METHOD = "get_ray_serve_publish_provider"
+
+
+class PublishProvider(ABC):
+    @staticmethod
+    @abstractmethod
+    def publish(
+        config: Dict,
+        *,
+        name: str,
+        ray_version: str,
+        base_image: Optional[str] = None,
+    ):
+        pass
+
+
+def get_publish_provider(provider_name: str) -> PublishProvider:
+    publish_provider = {
+        "anyscale": AnyscalePublishProvider,
+    }.get(provider_name, None)
+
+    if publish_provider is None:
+        try:
+            publish_provider_factory = import_attr(
+                f"{provider_name}.{PUBLISH_PROVIDER_FACTORY_METHOD}"
+            )
+        except (ModuleNotFoundError, AttributeError):
+            raise ModuleNotFoundError(
+                f"Failed to import '{PUBLISH_PROVIDER_FACTORY_METHOD}' "
+                f"from publish provider module '{provider_name}'."
+            )
+        publish_provider = publish_provider_factory()
+
+    return publish_provider
+
+
+class AnyscalePublishProvider(PublishProvider):
+    def publish(
+        config: Dict,
+        *,
+        name: str,
+        ray_version: str,
+        base_image: Optional[str] = None,
+    ):
+        service_config = {
+            "cloud": "anyscale_v2_default_cloud",
+            "name": name,
+            "ray_serve_config": config,
+        }
+        if base_image is not None:
+            service_config["cluster_env"] = base_image
+
+        # TODO(edoakes): use the Anyscale SDK (or another fixed entrypoint) instead of
+        # subprocessing out to the CLI.
+        with NamedTemporaryFile(mode="w") as f:
+            yaml.dump(service_config, f, default_flow_style=False)
+            f.flush()
+
+            try:
+                check_output(["anyscale", "service", "rollout", "-f", f.name])
```

Comment:
I thought about it, but:

- it'll basically amount to unit testing the `yaml` library so I think it might not add much value
- it'll take some annoying harness construction to fake out the subprocess call

I think it makes more sense to add tests when I remove the CLI subprocessing and replace it with SDK (making a fake SDK should be straightforward and useful).